9 спринт, первая попытка

muskilon opened the pull request.

Commit:
Добавил шрифты, вынес в стили часть верстки

Commit:
Сделал ночную тему, добавил цвета и изменил их названия в соответствии с Фигмой. Остался Switch

Commit:
Завершил ДЗ "Стили и темы"

Commit:
Спринт 9/25 → Тема 3/5: EditText и TextWatcher → Урок 3/3 ДЗ

Commit:
Спринт 9 финал. 1 попытка.

Commit:
Спринт 9 финал. 2 попытка.

muskilon commented:
Добавил шрифты

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -1,18 +1,63 @@
         val backArrow = findViewById<ImageView>(R.id.arrow_back)
         backArrow.setOnClickListener {
             val back = Intent(this, MainActivity::class.java)
             startActivity(back)
```
Для перехода назад надо использовать не интент, а методы finish() или onBackPressed(), потому что иначе вместо возврата в предыдущий экран создается новый экземпляр MainActivity и при возврате обратно на главный экран и нажатии на кнопку телефона "назад" мы "вернемся" снова на этот экран. На эту тему можно почитать про стек активити.

muskilon replied:
done

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -1,18 +1,63 @@
+    }
+    companion object {
+        const val INPUT_STRING = "INPUT_STRING"
+        const val SEARCH_INPUT_DEF = ""
```
Пустую строку выносить в константу - лишнее

muskilon replied:
done

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -1,18 +1,63 @@
+        outState.putString(INPUT_STRING, searchInput)
+    }
+    companion object {
+        const val INPUT_STRING = "INPUT_STRING"
```
private

muskilon replied:
done

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -1,18 +1,63 @@
+        }
+        textInputEdit.addTextChangedListener(simpleTextWatcher)
+    }
+    private var searchInput: String = SEARCH_INPUT_DEF
```
Переменные объявляют в начале класса обычно

muskilon replied:
done

faritowich commented on app/src/main/java/com/example/playlistmaker/SettingsActivity.kt:
```diff
@@ -1,21 +1,49 @@
         val backArrow = findViewById<ImageView>(R.id.arrow_back)
         backArrow.setOnClickListener {
             val back = Intent(this, MainActivity::class.java)
             startActivity(back)
```
См. коммент в SearchActivity

muskilon replied:
done(отдельное спасибо, у меня заработала корректно смена темы!)

faritowich commented on app/src/main/java/com/example/playlistmaker/SettingsActivity.kt:
```diff
@@ -1,21 +1,49 @@
+            }
+            val shareApp = Intent.createChooser(sendApp, null)
+            startActivity(shareApp)
+        }
```
Можно упростить до
```
            Intent().apply {
                action = Intent.ACTION_SEND
                putExtra(Intent.EXTRA_TEXT, getString(R.string.share_app_url))
                type = "text/plain"
                Intent.createChooser(this, null)
            }
```

muskilon replied:
надеюсь, правильно понял. done

faritowich commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -7,28 +7,58 @@
-            android:textSize="@dimen/font_size_medium"
-            android:textColor="@color/menu_black"
-            android:paddingLeft="12dp"
+            style="@style/NavigationToolBarText"
```
LinearLayout c иконкой и текстом можно заменить на один TextView с атрибутом drawable
Можно также для тулбара использовать androidx.appcompat.widget.Toolbar

muskilon replied:
Если это не критичное замечание, то я бы хотел оставить его на потом, что бы вдумчиво разобраться с Toolbar, а сейчас время поджимает(

faritowich commented on app/src/main/res/values/dimens.xml:
```diff
@@ -7,5 +7,6 @@
     <dimen name="font_size_medium">22sp</dimen>
     <dimen name="font_size_small">16sp</dimen>
     <dimen name="settings_items_height">61sp</dimen>
```
sp используется для размеров текста. Для остальных размеров - dp: https://developer.android.com/training/multiscreen/screendensities

muskilon replied:
done

faritowich commented on app/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <style name="MainMenuButtons">
```
Стиль кнопок на главном экране должен наследоваться от Widget.MaterialComponents.Button.UnelevatedButton.

muskilon replied:
done

faritowich commented on app/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,74 @@
+    <style name="MainMenuButtons">
+        <item name="android:layout_width">match_parent</item>
+        <item name="android:layout_height">0dp</item>
+        <item name="android:layout_marginBottom">6dp</item>
```
Лучше внешние отступы, высоту и ширину, веса, гравити не выносить в стиль. Если на экране будет много элементов, будет сложнее ориентироваться с их взаимным расположением, когда отвечающие за это атрибуты разнесены по разным файлам - что-то может быть в лейауте, что-то в стиле. В стилях обычно содержат атрибуты, отвечающие за внешний вид вьюхи и вообще только за то, что находится внутри нее - цвета, шрифты, фоны и т.д. Бывают исключения, когда, например, мы точно знаем, что у кнопки всегда должна строго определенная высота и ширина - тогда да, имеет смысл выносить

muskilon replied:
Согласен, но вот конкретно в этом случае, мы, как раз, знаем точно, что кнопки одинаковые. И при добавлении родительского стиля было очень удобно, что все вынесено в стили. Если не против, то я не хотел бы возвращать разметку в лейаут.

Commit:
Спринт 9 финал. 3 попытка.
Исправил интент с отправкой письма, по остальным правкам отписался в комментариях. Внезапная проблема в студии с отображением switcher, однако и в эмуляторе, и на физическом устройстве все в порядке. Так и не понял в чем проблема.

Commit:
10 спринт. Все работает. Остался неуточненный вопрос по поводу длинны текста в названиях. Пока реализовано просто ограничением на количество строк в TextView
